fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java

```diff
@@ -62,85 +63,90 @@ public List<ColumnRefOperator> getOutputExpressions() {
+            scanOperator.setPredicate(newPredicate);
         }
+        return true;
+    }
 
-        Set<Long> outdatePartitionIds = Sets.newHashSet();
-        for (Partition partition : mv.getPartitions()) {
-            if (partitionsToRefresh.contains(partition.getName())) {
-                outdatePartitionIds.add(partition.getId());
+    private List<Range<PartitionKey>> getUptodatePartitionRangeForTable(OlapTable partitionByTable,
+                                                                        MaterializedView mv,
+                                                                        Set<String> mvPartitionNamesToRefresh) {
+        Set<String> modifiedPartitionNames = mv.getUpdatedPartitionNamesOfTable(partitionByTable);
+        List<Range<PartitionKey>> baseTableRanges = getUptodatePartitionRange(partitionByTable, modifiedPartitionNames);
+        List<Range<PartitionKey>> mvRanges = getUptodatePartitionRange(mv, mvPartitionNamesToRefresh);
+        List<Range<PartitionKey>> uptodateBaseTableRanges = Lists.newArrayList();
+        for (Range<PartitionKey> range : baseTableRanges) {
+            if (mvRanges.stream().anyMatch(mvRange -> mvRange.encloses(range))) {
+                uptodateBaseTableRanges.add(range);
             }
         }
-        if (outdatePartitionIds.size() == mv.getPartitions().size()) {
-            // all partitions are out of date
-            // should not reach here, it will be filtered when registering mv
-            return;
+        return uptodateBaseTableRanges;
+    }
+
+    private List<Range<PartitionKey>> getUptodatePartitionRange(OlapTable table, Set<String> modifiedPartitionNames) {
+        List<Range<PartitionKey>> resultRanges = Lists.newArrayList();
+        // partitions that will be excluded
+        Set<Long> modifiedIds = Sets.newHashSet();
+        for (Partition p : table.getPartitions()) {
+            if (modifiedPartitionNames.contains(p.getName())) {
+                modifiedIds.add(p.getId());
+            }
         }
-        // now only one column is supported
-        Column partitionColumn = exprPartitionInfo.getPartitionColumns().get(0);
-        List<Range<PartitionKey>> uptodatePartitionRanges = exprPartitionInfo.getRangeList(outdatePartitionIds, false);
+        RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) table.getPartitionInfo();
+        List<Range<PartitionKey>> uptodatePartitionRanges = rangePartitionInfo.getRangeList(modifiedIds, false);
         if (uptodatePartitionRanges.isEmpty()) {
-            return;
+            return resultRanges;
         }
-        List<Range<PartitionKey>> finalRanges = Lists.newArrayList();
+
         for (int i = 0; i < uptodatePartitionRanges.size(); i++) {
             Range<PartitionKey> currentRange = uptodatePartitionRanges.get(i);
             boolean merged = false;
-            for (int j = 0; j < finalRanges.size(); j++) {
+            for (int j = 0; j < resultRanges.size(); j++) {
```

Comment:
Is the `resultRanges` initialized ?

Reply:
no, it is inited to Lists.newArrayList()